### PR TITLE
Reshape using row major layout

### PR DESCRIPTION
**Pull request by mtopalovicTT**
`reshape` currently doesn't work properly in `tile` major layout, thus we want to
convert input tensor to `row` major layout and to execute reshape.
Implemented solution uses pass which adds necessary ops which are needed in order to convert
input tensor from `tile` to `row` and back.

1. We create `empty` op which will have the same mem_layout/mem_space and
   grid size as original input tensor
2. We call `to_layout` and we pass input tensor as first operand and empty
   op from 1 as second operand (dps).
3. We create new empty tensor which will be used as dps operand for
  `reshape`
4. We invoke reshape with first operand from 2 and second operand 3
   (dps).
5. We invoke `to_layout` again on `reshape` result and we pass in original dps operand.

Other approach was to change reshape op in mlir rt, and convert input tensor to row layout if needed.
This approach is more clear (see b409044a792bbfc5fee8fb4f47fdb0004c0f539c) but downside is 
it doesn't cover emitc.

One more thing I tried is to squeeze changes into ttir->ttnn lowering of `reshape` op but this meant
I had to copy bunch of code from different places (toLayout lowering/empty op). Also it didn't make
sense to me to create ttnn ops (like empty/tolayout) which have their ttir equivalents inside `reshape` rewriter.

Misc:
* Fixed bug where we added ToDevice op in Layout lowering for tensors which are already on device

### CODE REVIEW

**Comment by jnie-TT** [comment]
@mtopalovicTT would it make sense to convert to row major layout only if the last two dimensions are not multiples of tile dimensions? AFAIK ttnn reshape ops should work correctly if we have height and width dimensions that are divisble by the tile size.

**Comment by mtopalovicTT** [comment]
@jnie-TT Problem is reshape in tile on metal doesn't produce right results even if you supply padded tensor. You would have to be careful how you place data inside the tensor for reshape to work in tile. Check this issue: https://github.com/tenstorrent/tt-metal/issues/12866

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 1]
```diff
@@ -797,6 +819,75 @@ class TTIRLayout : public impl::TTIRLayoutBase<TTIRLayout> {
...
+                        tensorGrid, getDataType(inTy), getMemoryLayout(inTy));
+    return RankedTensorType::get(inTy.getShape(), inTy.getElementType(),
+                                 newLayout);
+  }
+
+  LogicalResult matchAndRewrite(ReshapeOp op,
+                                PatternRewriter &rewriter) const final {
+    RankedTensorType inTy = op.getInput().getType();
+    LayoutAttr inputLayout = mlir::cast<LayoutAttr>(inTy.getEncoding());
+    if (!isa<tt::TileType>(inputLayout.getMemref().getElementType())) {
```
Comment (rpavlovicTT) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:841]: LayoutAttr has `isTiled()` function, you should use it

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 2]
```diff
@@ -797,6 +819,75 @@ class TTIRLayout : public impl::TTIRLayoutBase<TTIRLayout> {
...
+        op.getOperandConstraints());
+
+    rewriter.replaceOpWithNewOp<ttir::ToLayoutOp>(op, op.getResult().getType(),
+                                                  newReshapeOp.getResult(),
+                                                  op.getOperands().back());
+
+    return success();
+  }
+
+  const TypeConverter *converter;
```
Comment (rpavlovicTT) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:870]: Unused

File: test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline="system-desc-path=%system_desc_path%" %s > %t.mlir
+// RUN: FileCheck %s --input-file=%t.mlir
+// RUN: ttmlir-translate --ttnn-to-flatbuffer %t.mlir > %t.ttnn
+#any_device_tile = #tt.operand_constraint<dram|l1|tile|any_device_tile>
+
```
Comment (rpavlovicTT) [inline test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir:5]: Should you add some tests that start with tiled tensors? 
Also, add some CHECKs for layout ops (whatever they become in TTNN).

Comment (mtopalovicTT) [reply test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir:5]: I've added two more tests in silicon which have start with tiled tensor.
> Also, add some CHECKs for layout ops (whatever they become in TTNN).
Added this to simple_reshape.mlir in Dialect directory. It checks exact ordering of ops and input values.

Comment (rpavlovicTT) [reply test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir:5]: >I've added two more tests in silicon which have start with tiled tensor.

Maybe I'm missing some default behaviour, but how do you know they start with tiled tensor? There is no Layout attribute defined in the mlir file that specifies tiled layout.

Comment (mtopalovicTT) [reply test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir:5]: #any_device_tile = #tt.operand_constraint<dram|l1|**tile**|any_device_tile>
Should be this right?

**Review by nvukobratTT — COMMENTED** [review]
@mtopalovicTT is this a workaround for the below issue? Or do we plan to push effort on TTNN to support this behavior as well? 
- https://github.com/tenstorrent/tt-metal/issues/12866

On TTNN issue, I see that they're mentioned to have this working by EOW. Can you just explain in a bit more details if I'm missing some points here :))

**Comment by mtopalovicTT** [comment]
@nvukobratTT We wanted to have some workaround until they fix reshape, we didn't know timeline ahead of time...
If they are sure it's EOW and there is nothing blocking uplifting metal version we don't need this PR. On other side if we merge, rolling this back would be trivial.

File: test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir (hunk 3 above)
Comment (rpavlovicTT) [reply test/ttmlir/Silicon/TTNN/reshape/reshape_tests.mlir:5]: Got it, missed `tile` part of the constraint. Thanks!

**Review by rpavlovicTT — APPROVED** [review]

**Comment by nvukobratTT** [comment]
> @nvukobratTT We wanted to have some workaround until they fix reshape, we didn't know timeline ahead of time... If they are sure it's EOW and there is nothing blocking uplifting metal version we don't need this PR. On other side if we merge, rolling this back would be trivial.

Thanks for the details! is it okay to wait until Monday for TTNN feedback? If this will take more time, we can probably go with workaround, but please create issues to revert the change latter if needed. 

@tt-mpantic let us know if you have more details around progress on this reshape, or should we ping them once again :))
- https://github.com/tenstorrent/tt-metal/issues/12866

**Comment by topke011** [comment]
Its fine with me.

**Comment by nvukobratTT** [comment]
> Its fine with me.

Seems like there is open PR for it :)) 🤘

**Comment by mtopalovicTT** [comment]
Okay, looks like they merged the change, I've added item to test and uplift metal version. Once everything is good I will close this PR.

**Comment by mtopalovicTT** [comment]
Closing this since reshape should be fixed with metal uplift that joca is driving with #870

**Pull request closed without merging**